### PR TITLE
zcash_client_sqlite crate

### LINKED ISSUE #71: Add feature flag to toggle between mainnet and testnet
Instead of requiring all APIs to contain parameter fields, we can leverage the Rust features system to have this handled at build time. The will be a feature on the `zcash_client_backend` crate and possibly on other crates depending on what makes sense.

### PR DESCRIPTION
**Pull request by str4d**
My [`note-spending-v7`](https://github.com/str4d/librustzcash/tree/note-spending-v7) branch is pretty stable at this point, and now being depended on more widely (after the ECC wallet was recently open-sourced). I rebased my branch on current master to fix merge conflicts (due to recent refactors), but it's otherwise unaltered.

Closes #71.

### CODE REVIEW

**Comment by str4d** [comment]
@gmale that commit was separately applied in #236; rebasing eliminates commits that already exist in master (as they would otherwise be empty).

File: zcash_client_sqlite/src/init.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+///
+/// ```
+/// use tempfile::NamedTempFile;
+/// use zcash_client_sqlite::init::init_data_database;
+///
+/// let data_file = NamedTempFile::new().unwrap();
+/// let db_data = data_file.path();
+/// init_data_database(&db_data).unwrap();
+/// ```
+pub fn init_data_database<P: AsRef<Path>>(db_data: P) -> Result<(), Error> {
```
Comment (daira) [inline zcash_client_sqlite/src/init.rs:50]: This should require specifying privacy-relevant policy such as whether `ovk`s are stored.

Comment (str4d) [reply zcash_client_sqlite/src/init.rs:50]: Alternatively (or in addition), `transact::create_to_address` could require specifying a per-transaction policy; this is where the policy would be enforced.

Comment (str4d) [reply zcash_client_sqlite/src/init.rs:50]: Specifically, here is where the current default policy is implemented: https://github.com/zcash/librustzcash/pull/246/files#diff-0db70545aba62e450ec1e1634b293f05R104

**Comment by daira** [comment]
There should also be an API to delete data on transactions held by the wallet, e.g. `ovk`s and other viewing keys, and any data cached from them. As @str4d said on Discord:

> Without ovk you lose the non-change recipients and memos [if you restore from the backup seed]. You still learn how much was spent and when.

The use case here is that a wallet can have a button that someone might choose to use of they are going into a potentially coercive environment where data on their device could be seized.

**Comment by imichaelmiers** [comment]
I agree with all of this, but we should be clear on which does what.  I could easily see down stream devs confusing what does what. This might result in end users deleting history but not OVK and thinking they've erased a past sensitive purchase. Or  that deleting OVK erases past purchases.

Ideally there should be some policy switch with clear labels:
1) no history 
2) (if we support this) device lcoal history file  but no on chain storage. 
3) history stored on chain encrypted and undeletable , but only accessible with off device backup key. Erasing this key may lose access to money
4) history stored on chain encrypted and undeletable,  accessible from keys stored on  the device.
Erasing this key may lose access to money

**Comment by gmale** [comment]
This all touches on a much larger issue of whether librustzcash should store any state. I would propose that we **not** bandaid key storage and instead take a thoughtful approach to implementing https://github.com/zcash/librustzcash/issues/142 which returns control of data storage to the wallet.

> I agree with all of this, but we should be clear on which does what. 

I strongly agree with that statement. We seem to be discussing features and decisions that probably should be made at the wallet level, rather than within librustzcash. The root issue seems to be that, with this PR, `librustzcash` will officially manage state and we could avoid that entirely and let wallets manage state, instead.

**Comment by gmale** [comment]
The discussion here relates to the ["Hide Transaction" feature request](https://github.com/zcash/zcash-android-wallet/issues/173) on the wallet.

**Comment by str4d** [comment]
> Blocking request to require specifying whether `ovk`s are stored when opening the database.

I've implemented an outgoing viewing key usage policy for `transact::create_to_transaction`, which provides per-transaction granularity. Previous behaviour can be retained by using `OvkPolicy::Sender`.

> This will require a version bump according to semver.

`zcash_client_sqlite` is internally still version 0.0.0, and has thus far only been used by pinning a git commit. The version bump will occur inherently when 0.1.0 is released.

> There should also be an API to delete data on transactions held by the wallet, e.g. `ovk`s and other viewing keys, and any data cached from them.

Let's implement this in a subsequent PR.

File: zcash_client_sqlite/src/transact.rs [hunk 2]
```diff
@@ -0,0 +1,848 @@
...
+    let id_tx = data.last_insert_rowid();
+
+    // Mark notes as spent.
+    //
+    // This locks the notes so they aren't selected again by a subsequent call to
+    // create_to_address() before this transaction has been mined (at which point the notes
+    // get re-marked as spent).
+    //
+    // Assumes that create_to_address() will never be called in parallel, which is a
+    // reasonable assumption for a light client such as a mobile phone.
```
Comment (daira) [inline zcash_client_sqlite/src/transact.rs:330]: This assumption should be documented in the function comment.

File: zcash_client_sqlite/src/query.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+    match Amount::from_i64(balance) {
+        Ok(amount) if !amount.is_negative() => Ok(amount),
+        _ => Err(Error(ErrorKind::CorruptedData(
+            "Sum of values in received_notes is out of range",
+        ))),
+    }
+}
+
+/// Returns the memo for a received note, if it is known and a valid UTF-8 string.
```
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_client_sqlite/src/query.rs:94]: The way this crate is designed (specifically that `db_data` is read-only outside the crate), it's expected that the caller can query the database for that information themselves.

File: zcash_client_sqlite/src/init.rs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            tx INTEGER NOT NULL,
+            output_index INTEGER NOT NULL,
+            account INTEGER NOT NULL,
+            diversifier BLOB NOT NULL,
+            value INTEGER NOT NULL,
+            rcm BLOB NOT NULL,
+            nf BLOB NOT NULL UNIQUE,
+            is_change INTEGER NOT NULL,
+            memo BLOB,
+            spent INTEGER,
```
[the remark this answers is not in the document]
Comment (str4d) [reply zcash_client_sqlite/src/init.rs:94]: No, this is a foreign key to the `transactions` database, pointing at the transaction this note is spent in.

File: zcash_client_sqlite/src/transact.rs (hunk 2 above)
Comment (str4d) [reply zcash_client_sqlite/src/transact.rs:330]: It is:
> Do not call this multiple times in parallel, or you will generate transactions that double-spend the same notes.

**Pull request merged**